### ISSUE TITLE
CC: Add ability to set offline_fs_keys/resources

### ISSUE THREAD

**Issue by stevenhorsman** [kata-containers/kata-containers#8180]
In many of the `CC` tests, we use the offline_fs_kbc to run tests that depend on attestation measurement 'released' resources and keys to run on both the SecureExecution platform (which has an encrypted image to ensure the local filesystem can be trusted) and in non-TEE environments for testing and so that users without access to expensive TEE hardware can see try out and develop on the platform.

To set up these files we need to have the ability to write the [/etc/aa-offline_fs_kbc-keys.json](https://github.com/confidential-containers/guest-components/blob/b1e12bef36e105e0aa7b3cef683ee67cd51e3310/attestation-agent/kbc/src/offline_fs_kbc/mod.rs#L21) and [/etc/aa-offline_fs_kbc-resources.json](https://github.com/confidential-containers/guest-components/blob/b1e12bef36e105e0aa7b3cef683ee67cd51e3310/attestation-agent/kbc/src/offline_fs_kbc/mod.rs#L22C31-L22C68) on the guest images (and sometime with some supporting files to that the resources link to).

In `CCv0` this is currently done in most of the tests using the [`cp_to_guest_img`](https://github.com/kata-containers/tests/blob/6cf430642335c010014b5d9355a861328c4213ed/lib/common.bash#L295), util function, which mounts the guest image, adds the extra content over and then compresses it back up. This is not ideal as it involves up modifying the guest image and on certain platforms, such as AKS where the kata-deploy is used to install kata containers, getting access to the worker to do these updates might prove to be very difficult.

An alternative approach could be to use [policies](https://github.com/orgs/kata-containers/projects/39/views/1) as a vehicle to pass through this configuration to the guest, potentially as an annotation as part of the pod config. However I'm not sure that polcies are fully adopted and tested on all platforms at the moment, so there might be some foundational work that is required first.

**Comment by fitzthum** [comment]
Another option might be to revive the `sample` KBC and either hardcode test keys into it or have it pull test credentials from the network somewhere.

I'm not sure that policy will be able to solve this problem, at least in the short term. At the moment the policy support we have is limited to validating the Kata Agent API.

**Comment by danmihai1** [comment]
@stevenhorsman @fitzthum , adding ability to host JSON format data in a policy document is next on my TODO list - I will let you know when it's working/tested (in the main branch). 

In fact hosting such JSON data in the policy file already works - we just don't have yet a well-established way to retrieve that data from OPA.

**Comment by danmihai1** [comment]
@stevenhorsman @fitzthum I added some examples of sending JSON data from the Host to the Guest using a Policy file, in https://github.com/kata-containers/kata-containers/pull/8230/files

@stevenhorsman , when you get a chance, please explain which Guest module is expected to read this data - to make sure we have a good plan for querying it.

Thanks!

**Comment by stevenhorsman** [comment]
Thanks for taking a look at this Dan. So the current approarch is that the attestation-agent and confidential-data-hub will be reading the [/etc/aa-offline_fs_kbc-keys.json](https://github.com/confidential-containers/guest-components/blob/b1e12bef36e105e0aa7b3cef683ee67cd51e3310/attestation-agent/kbc/src/offline_fs_kbc/mod.rs#L21) and [/etc/aa-offline_fs_kbc-resources.json](https://github.com/confidential-containers/guest-components/blob/b1e12bef36e105e0aa7b3cef683ee67cd51e3310/attestation-agent/kbc/src/offline_fs_kbc/mod.rs#L22C31-L22C68) as part of the offline_fs_kbc module.

So based on this there are potentially a couple of things we could take (that occur to me now, I'm sure others have better ideas)
1. We update the offline_fs_kbc, or probably create a new one that is able to query OPA for the config. I kinda like this, but it's not backwards compatible and required the syncing of kata-agent and guest-components code which could be fragile
2. We write something in the kata-agent, that related to start-up, support would query OPA and get the specific config for the offline_fs_kbc (if supplied) and write it out to `/etc/aa-offline_fs_kbc-keys.json` and `/etc/aa-offline_fs_kbc-resources.json` on the guest system. This is an extra step, but decouples things better, so probably the best approach I can think of for this at the moment.

**Comment by fitzthum** [comment]
Keep in mind that the policy is measured but not encrypted. I guess that could be ok for some resources, but I don't really see it as a solution for delivering secrets. Is this just for testing? If so, I'm not sure it makes sense to add a codepath for tests that we don't use on other platforms. Note that s390x, which relies on the `offline_fs_kbc` doesn't seem to have any way to verify the policy.

Like I said, I think the most robust test would be to create a sample evidence provider and do a remote attestation.

**Comment by danmihai1** [comment]
@stevenhorsman , I think we should encourage the features consuming these JSON file to retrieve the same data using an API, that:

1. Reads that data from the Policy - as in the https://github.com/kata-containers/kata-containers/pull/8261 example.
2. Optionally, if the JSON file is present, could read that data from the JSON file.

But, if we'll really need the Kata Agent to create JSON files from Policy, that should take approximately 10 lines of code - we can work on that together if you want.

We can also add support for encrypted Policy if some of the use cases need that.

**Comment by ariel-adam** [comment]
@stevenhorsman going back to this thread from November.
What should be the next steps and who should own it?

**Comment by stevenhorsman** [comment]
Dan might be able to correct me as I've not been involved in the calls, but I think this is blocked by the decision made about initdata - when I wrote it I assumed we'd be using policy to provide config, but I'm not sure where that landed

**Comment by ariel-adam** [comment]
@danmihai1 should we move it to blocked?

**Comment by danmihai1** [comment]
@ariel-adam, arbitrary data can be sent to the Guest using the current Agent Policy. We also have the mechanisms in place to verify the integrity of that data through remote attestation on platforms that support such verification (I am aware of that kind of remote attestation feature being available for SEV-SNP and TDX).

I know there are folks who would like to change the format of this data (from OPA/Rego to JSON). I disagree with that change - but we'll have to decide on that as a community. In any case, even if we change from OPA/Rego to JSON, sending arbitrary Policy data to the Guest will continue to work.

Maybe there are missing features for additional use cases and/or additional hardware platforms. I am not the right person to push for these additional features, because I am not familiar with these use cases. So, perhaps additional features and/or changes to Agent Policy should be tracked using different github issues. From my point of view, the features originally described in this issue are available already - although not yet using JSON as initially described above. They are available by using Rego.

**Comment by fitzthum** [comment]
I would agree with closing the issue, but I think the solution is only partly what @danmihai1 describes. The InitData/policy is how we will be provisioning configuration data but this issue also mentions keys (i.e. secrets), which we can't deliver via that (unencrypted) method. For that I think the answer is more that we are going to use the cc-kbc with the sample attester for most tests. I think this addresses this will address the original concerns. See https://github.com/confidential-containers/confidential-containers/issues/184 for more details.

**Comment by stevenhorsman** [comment]
Based on the comments above closing this.
